pkg/storage/etcd/etcd_watcher.go

```diff
@@ -144,6 +144,18 @@ func (w *etcdWatcher) etcdWatch(ctx context.Context, client etcd.KeysAPI, key st
 	defer util.HandleCrash()
 	defer close(w.etcdError)
 	defer close(w.etcdIncoming)
+
+	// We need to be prepared, that Stop() can be called at any time.
+	// It can potentially also be called, even before this function is called.
+	// If that is the case, we simply skip all the code here.
+	w.stopLock.Lock()
+	if w.stopped {
+		// Watcher has already been stopped - don't event initiate it here.
+		return
```

Comment:
If the higher level code called Stop and then re-initiated a watch, that's pretty disturbing behavior.